topic-operator/src/main/java/io/strimzi/operator/topic/Session.java

```diff
@@ -133,72 +136,102 @@ public void init(Vertx vertx, Context context) {
+            vertx.cancelTimer(this.timerId);
         }
-        vertx.executeBlocking(blockingResult -> {
-            long timeout = 120_000L;
-            long deadline = System.currentTimeMillis() + timeout;
-            LOGGER.info("Stopping");
-            LOGGER.debug("Stopping kube watch");
-            topicWatch.close();
-            LOGGER.debug("Stopping zk watches");
-            topicsWatcher.stop();
-
-            Promise<Void> promise = Promise.promise();
-            Handler<Long> longHandler = new Handler<>() {
-                @Override
-                public void handle(Long inflightTimerId) {
-                    if (!topicOperator.isWorkInflight()) {
-                        LOGGER.debug("Inflight work has finished");
-                        promise.complete();
-                    } else if (System.currentTimeMillis() > deadline) {
-                        LOGGER.error("Timeout waiting for inflight work to finish");
-                        promise.complete();
-                    } else {
-                        LOGGER.debug("Waiting for inflight work to finish");
-                        vertx.setTimer(1_000, this);
+        ExecutorService shutdownExecutor = Executors.newSingleThreadExecutor();
+        closeKubernetesWatch()
+                .onSuccess(ignored -> LOGGER.debug("Kubernetes watch closed"))
+                .onFailure(cause -> LOGGER.error("Failed to close Kubernetes watch", cause))
+            .compose(ignored -> stopZooKeeperWatcher())
+                .onSuccess(ignored -> LOGGER.debug("ZooKeeper topics watcher stopped"))
+                .onFailure(cause -> LOGGER.error("Failed to stop ZooKeeper topics watcher", cause))
+            .compose(ignored -> stopTopicStoreService())
+                .onSuccess(ignored -> LOGGER.debug("Topic store service stopped"))
+                .onFailure(cause -> LOGGER.error("Failed to stop topic store service", cause))
+            .compose(ignored -> disconnectFromZk(shutdownExecutor))
+                .onSuccess(ignored -> LOGGER.debug("ZooKeeper disconnected"))
+                .onFailure(cause -> LOGGER.error("Failed to disconnect from ZooKeeper", cause))
+            .compose(ignored -> closeKafkaAdminClient(shutdownExecutor))
+                .onSuccess(ignored -> LOGGER.debug("Kafka admin client closed"))
+                .onFailure(cause -> LOGGER.error("Failed to close Kafka admin client", cause))
+            .compose(ignored -> closeHttpHealthServer())
+                .onSuccess(ignored -> LOGGER.debug("HTTP health server closed"))
+                .onFailure(cause -> LOGGER.error("Failed to close HTTP health server", cause))
+            .compose(ignored -> {
+                try {
+                    shutdownExecutor.shutdown();
+                    shutdownExecutor.awaitTermination(10, TimeUnit.SECONDS);
+                } catch (InterruptedException e) {
+                    if (!shutdownExecutor.isTerminated()) {
```

Comment:
It's worth adding a comment to explain why this behaviour for handling the InterruptedException is thought to be correct.